### ISSUE TITLE
Types Statements

### ISSUE THREAD

**Issue by ije** [alephjs/alef-component-rfcs#9]
> Alef Component use types statements to specify some features.

## Props

```jsx
const n: Prop<number> = 1
```

## Slots

```jsx
const children: Prop<Children> = null
```

## Use Context

```jsx
const n: Context<string> = name
```

**Comment by shadowtime2000** [comment]
@ije For the Ref example you have it resetting a `const` variable, which will basically break. Maybe this could be the `Prop<T>` type:
```ts
type Prop<T> = undefined | null | T
```

**Comment by ije** [comment]
@shadowtime2000  make sense, but people may want set a default value when the prop or context is undefined, i prefer `<string | null>` if you don't want to use an initial value:

```jsx
let name: Prop<string | null> = null
```

**Comment by shadowtime2000** [comment]
I am a little confused, would this make the compiler recognize what the value is, memo, context, ref, etc based on their types, basically forcing usage of some static type checker?

**Comment by ije** [comment]
@shadowtime2000 yes, `AlefType<T>` will tell compiler the variable whether it is a reactive role explicitly.

**Comment by ije** [comment]
<img width="828" alt="Screen Shot 2020-12-13 at 3 22 23 PM" src="https://user-images.githubusercontent.com/2883484/102005840-0cbb9080-3d57-11eb-9981-4d58dd34440b.png">
